### ISSUE TITLE
Question/Proposal: Passphrase protected Master Private Key?

### ISSUE THREAD

**Issue by Anthonyhawkins** [saltstack/salt#35303]
### Description of Issue/Question

We are currently in the process of deploying salt into our environment.  One of our security policies dictates that all private-keys are passphrase protected.

I have taken a look at the code and current functionality of the salt-master and it does not appear to be supported.  We'd like to develop this feature and contribute it back, but first we have some questions.

- **Would this actually be accepted?:** Our main concern is ensuring that option of having a passphrase protected private key for the master is in-line with what is envisioned for salt.  i.e. Is this something that would be added if the implementation was solid etc.  or, is the demand not enough that it wouldn't be added regardless.

- **Where would this actually be implemented?:** From what I can see the overall low-laying function call which needs to be modified is in salt/crypt.py where it makes a call out to RSA object's export method: https://www.dlitz.net/software/pycrypto/api/current/Crypto.PublicKey.RSA._RSAobj-class.html#exportKey . From there, we would bubble-up that parameter to the salt-master CLI with an option for --key-pass=<passphrase> etc. as an optional requirement if the key is passphrase protected or prompt the user before the process starts.

### Some Points regarding the environment:
- Salt Master is running as non-root
- Minions are running as root

As always, thanks for the feedback and assistance.

**Comment by damon-atkins** [comment]
Not sure how this is different to SSH.   I believe the keys are for verification.  If this did go ahead every-time the master is started it would require the passphrase being entered.  

What I am asking is, what are you trying to protect from?  A person with sudo access to root?  Preventing someone from signing keys/accepting servers?  Someone copying the cert and using it in some way?

**Comment by Ch3LL** [comment]
ping @thatch45 can I get your input on this issue in regards to @Anthonyhawkins 's proposal to adding the ability use passwords on keys and his question about where to fix it? Thanks

**Comment by thatch45** [comment]
Would this actually be accepted? - YES! I would like to see this feature in Salt, I think it would be well received.
Where would this actually be implemented?
Yes, crypt.py would be a good place, but we would also need to add support for this inside salt-key. I think that the main challenge would be how to read in the key, since you would not want the password to be stored on disk, and you also do not want the password to be entered in the startup command, since that would save the password in the shell history as well as show up in the process table. This means that the password would need to be read at a startup prompt for the master, or accepted via stdin. We don't have facilities to do this right now, although they can be easy enough to implement.

So long story short, people have expressed a desire for this, just not enough of one to motivate the creation of the interface, so I would be happy to accept it, if it is well written,

@damon-atkins password protected keys are just keys that are encrypted (basically) using the password as the encryption key. This allows the private data to be stored at rest securely, and enables you the time to replace lost keys before they are encrypted. The standard is to use AES encryption on the key and there are likely routines in pycrypto to do it. As usual we will will not accept cryptographic code which does not use an upstream crypto lib like openssl (pycrypto is a wrapper around openssl)

**Comment by Anthonyhawkins** [comment]
Thanks all, for the quick response. 

Historically what we have done is anytime we need a password is to reach out to a custom vaulting utility which houses the password.  This is done via a command i.e.

```
some_process_which_needs_a_password --password=`my_password_getter_command --alias=admin-pass` 
```
This would return the password to stdout, however, this does not solve for having the password visible in the process output.  

One possible solution for this would be to not tell salt *what* the password is, but rather to tell *how* it can get the password.  

An example of how this would be done can be seen in the Apache HTTPd directive
```
SSLPassPhraseDialog exec:/path/to/passphrase-script
```

**Comment by thatch45** [comment]
We have a solution for that, if the password can be loaded from the config file then we can use sdb to store it in an external vault

**Comment by Anthonyhawkins** [comment]
This is good stuff.  One less wheel to re-invent regarding retrieving it from an external vault!  

Just to confirm I'm assuming (correct me if I'm mistaken) that the salt-master on first start reads in the config prior to creating it's private key and on every boot after that it reads the config before loading it's key.

**Comment by thatch45** [comment]
Correct :)

**Comment by Anthonyhawkins** [comment]
A PR has been created to continue this discussion.
https://github.com/saltstack/salt/pull/35542
